src/Chainweb/Pact/Types.hs

```diff
@@ -370,15 +377,45 @@ defaultOnFatalError lf pex t = do
   where
     errMsg = pack (show pex) <> "\n" <> t
 
+type ModuleInitCache = M.Map BlockHeight ModuleCache
+
 data PactServiceState = PactServiceState
     { _psStateValidated :: !(Maybe BlockHeader)
-    , _psInitCache :: !ModuleCache
+    , _psInitCache :: !ModuleInitCache
     , _psParentHeader :: !ParentHeader
     , _psSpvSupport :: !SPVSupport
     }
 makeLenses ''PactServiceState
 
 
+_debugMC :: Text -> PactServiceM cas ()
+_debugMC t = do
+  mc <- fmap (fmap instr) <$> use psInitCache
+  liftIO $ print (t,mc)
+  where
+    instr (ModuleData{..},_) = preview (_MDModule . mHash) _mdModule
+
+-- | Look up an init cache that is stored before the height of the current parent header.
+getInitCache :: PactServiceM cas ModuleCache
+getInitCache = get >>= \PactServiceState{..} ->
+    case M.lookupLE (pbh _psParentHeader) _psInitCache of
+      Just (_,mc) -> return mc
+      Nothing -> return mempty
+  where
+    pbh = _blockHeight . _parentHeader
+
+-- | Update init cache at adjusted parent block height (APBH), unioning contents with
+-- any cache found at or before APBH. APBH is modulated by first argument, in practice
+-- it is 'id' for genesis and 'succ' for thereafter.
+updateInitCache :: (BlockHeight -> BlockHeight) -> ModuleCache -> PactServiceM cas ()
```

Comment:
Do you think that it perhaps could be better for the first argument here to be a boolean (or some ADT) that makes its purpose easier to investigate?

Reply:
Why not just drop it? It seems that it's always `id` when the parent block height is 0 and `succ` otherwise.